engine/sources/local/singleflight.go

```diff
@@ -0,0 +1,102 @@
+package local
+
+import (
+	"context"
+	"sync"
+)
+
+// SingleflightGroup is similar to sync.Singleflight but:
+// 1. Handles context cancellation (ctx provided to callback is only cancelled once no one is waiting on the result)
+// 2. Caches results until all returned CachedResults are released
+type SingleflightGroup[K comparable, V any] struct {
```

Comment:
Note for reviewers: this is the heart of all the tricky concurrency handling we needed here. We use paths in the filesystem being synced as keys and run operations on the paths in the callback to `Do`.

Part of the utility is that equivalent operations on the same path from different concurrent syncs are deduped and don't stomp over each other.

The fact that the results are cached helps with deduping and also allows the code to identify the client filesystem is changing during the sync and error out (see `verifyExpectedChange` and its usage). If the returned result was cached from a previous operation and it doesn't match what we expected to apply, we know something changed in the middle of the sync.

The other important part is the behavior where results are cached until all returned references to them are `.Release()`'d. This allows us to essentially "lock" individual paths as they are referenced in a given sync but release them when we're done.

Reply:
So, this is part of the `localFSSharedState`. So, multiple mounters of the big shared local state don't run over each other.

But just to confirm - that means all operations for this now need to go through this group to avoid accidentally running over each other? Not really a concern, just wanting to check my understanding, though it does mean that we wouldn't be able to have multiple dagger engines accessing local cache refs from the same store (though I guess there are significantly more complicated blockers for this anyways).

Reply:
Yeah I mean the behavior before this was to only have one mount at a time and create a brand new empty cache ref if all others were in use at the time of the sync, so now it's possible to use the same one in parallel within the same process. Wasn't attempting to make it safe across processes yet, but that'd be doable with the hardlink-farm approach (since it would become one big "content store" of individual files essentially and thus feasible to make concurrent safe w/ flock). This PR paves the way for that approach, but too big a change on top of a big change, so saving for down the line once needed.